programs/stake/src/stake_state.rs

```diff
@@ -1152,6 +1165,28 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
     }
 }
 
+fn get_info_if_mergeable(
+    stake_keyed_account: &KeyedAccount,
+    clock: &Clock,
+    stake_history: &StakeHistory,
+) -> Result<(Meta, Option<Stake>), InstructionError> {
+    match stake_keyed_account.state()? {
+        StakeState::Stake(meta, stake) => {
+            // stake must not be in a transient state
+            match stake.delegation.stake_activating_and_deactivating(
+                clock.epoch,
+                Some(stake_history),
+                true,
+            ) {
+                (0, _, _) => Ok((meta, None)),
```

Comment:
just making sure: this arm is for fully-deactivated stake and just-activated stake delegated immediately within the current epoch?

Reply:
Also, this could allow merging these with any `StakeState::Initialized` stakes with `mergeable` `Meta`. I wonder there could be any irregular state because of corner-case operation...

Reply:
Nailed it!  I was actually surprised to see that the original code supports this case.  At first I was using `(0, 0, 0)` (fully deactivated) here, but a couple tests broke, which revealed the legacy support for merge-during-activation-epoch.

I think the only corner cases are:
1. merge undelegated into delegated during activation-epoch - the way we do the calculations today, that's no different than having merged before delegating.  Perhaps this warrants finding a good place for a comment, in case we change the math one day.
1. merge delegated into undelegated during activation-epoch - this is just a single instruction synonym for merge and deactivate.  A bit of user confusion at worst

I'm going to think a bit more on this one and how it affects the other assumptions I made while making these changes today, now that my brain is fresh

Reply:
It does look like merging `Initialized` stakes into mergeable `Delegated` stakes was already supported.
I haven't thought of any reason why `Initialized` would behave any differently than fully deactivated stake.